File: web/src/shared/engine.ts
/**
 * Client port of internal/modules/forms/engine.
 *
 * This decides what the respondent sees as they answer. The server keeps the
 * authoritative copy and re-evaluates every submission -- trusting the client's
 * claim about which fields it displayed would let a caller submit answers for a
 * branch they never saw, and therefore never saw the consent text for.
 *
 * The two implementations are held together by testdata/golden.json, which both
 * read unchanged. A rule engine that disagrees between browser and server
 * produces a form that validates on screen and fails on submit, which a person
 * filling it in experiences as the form simply being broken.
 */

export type FieldID = string
export type PageID = string

export interface Option {
  id: string
  label: string
}

/**
 * Mirrors domain.Field in internal/modules/forms/domain/schema.go.
 *
 * Field names are the Go json tags, not what reads well here. `max_mb` was
 * `max_bytes` in an earlier draft of this file, which meant an upload limit read
 * as undefined -- a size cap that silently is not one.
 */
export interface Field {
  type: string
  label: string
  required?: boolean
  hidden?: boolean

  options?: Option[]

  pii?: string
  sensitive?: boolean
  /** The field used to recognise a data subject across submissions. */
  identifier?: boolean

  scale?: number
  accept?: string[]
  /** Megabytes, per the Go tag -- not bytes. */
  max_mb?: number
  multiline?: boolean
  /** One of the named formats in format.ts. Empty means any text. */
  format?: string
  /** Bounds: YYYY-MM-DD for a date, a number for the numeric formats. */
  min?: string
  max?: string
}

export interface Page {
  id: PageID
  title: string
  next?: PageID
  fields: FieldID[]
}

export interface Condition {
  op: string
  field: FieldID
  value?: unknown
}

export interface Action {
  action: string
  target: string
}

export interface Rule {
  id: string
  on_page: PageID
  when: Condition
  then?: Action[]
  else?: Action[]
}

export interface Purpose {
  code: string
  required?: boolean
  label?: string
}

export interface Schema {
  v: number
  pages: Page[]
  fields: Record<FieldID, Field>
  rules?: Rule[]
  consent?: { purposes?: Purpose[] }
}

export type Answers = Record<string, unknown>

export interface Result {
  /** Fields actually shown, in page then declaration order. */
  visible: FieldID[]
  /** The visible fields that must be answered. */
  required: FieldID[]
  /** Pages traversed, in order. */
  path: PageID[]
}

export class EngineError extends Error {}

/** Evaluate walks the form as the answers direct and reports what was shown. */
export function evaluate(schema: Schema, answers: Answers): Result {
  const result: Result = { visible: [], required: [], path: [] }
  if (schema.pages.length === 0) return result

  const byPage = new Map<PageID, Rule[]>()
  for (const rule of schema.rules ?? []) {
    const list = byPage.get(rule.on_page)
    if (list) list.push(rule)
    else byPage.set(rule.on_page, [rule])
  }

  let current: PageID = schema.pages[0]!.id
  const seen = new Set<PageID>()

  // Publishing rejects cyclic navigation, but a schema written before that check
  // existed -- or edited straight in the database -- must not be able to hang
  // the tab. The visited set bounds the walk regardless.
  for (let i = 0; i <= schema.pages.length; i++) {
    const page = schema.pages.find((p) => p.id === current)
    if (!page) throw new EngineError(`page "${current}" referenced but not defined`)
    if (seen.has(current)) throw new EngineError(`navigation cycle at page "${current}"`)
    seen.add(current)
    result.path.push(current)

    const { visible, required, next } = evalPage(schema, page, byPage.get(current) ?? [], answers)
    result.visible.push(...visible)
    result.required.push(...required)

    if (next === '') return result
    current = next
  }
  throw new EngineError(`navigation exceeded ${schema.pages.length} pages`)
}

interface PageResult {
  visible: FieldID[]
  required: FieldID[]
  /** '' means the form ends here. */
  next: PageID | ''
}

function evalPage(schema: Schema, page: Page, rules: Rule[], answers: Answers): PageResult {
  // Start from the declared state, then let rules amend it. A field marked
  // hidden appears only if a rule shows it.
  const shown = new Map<FieldID, boolean>()
  const must = new Map<FieldID, boolean>()
  for (const fid of page.fields) {
    const f = schema.fields[fid]
    if (!f) throw new EngineError(`page "${page.id}" lists undefined field "${fid}"`)
    shown.set(fid, !f.hidden)
    must.set(fid, Boolean(f.required))
  }

  let jump: PageID = ''
  let ended = false
  let decided = false

  // One pass, in declaration order. Later rules override earlier ones, which
  // makes the outcome depend only on the schema's own ordering.
  for (const rule of rules) {
    const match = evalCondition(rule.when, answers)
    const actions = (match ? rule.then : rule.else) ?? []
    for (const a of actions) {
      switch (a.action) {
        case 'show':
          shown.set(a.target, true)
          break
        case 'hide':
          shown.set(a.target, false)
          break
        case 'require':
          must.set(a.target, true)
          break
        case 'optional':
          must.set(a.target, false)
          break
        case 'goto':
          jump = a.target
          ended = false
          decided = true
          break
        case 'end':
          jump = ''
          ended = true
          decided = true
          break
        default:
          throw new EngineError(`rule "${rule.id}": unknown action "${a.action}"`)
      }
    }
  }

  const visible: FieldID[] = []
  const required: FieldID[] = []
  for (const fid of page.fields) {
    if (!shown.get(fid)) continue
    visible.push(fid)
    if (must.get(fid)) required.push(fid)
  }

  if (decided && ended) return { visible, required, next: '' }
  if (decided) return { visible, required, next: jump }
  return { visible, required, next: defaultNext(schema, page.id) }
}

/** Where the form goes when no rule decides: the declared next, or the following
 *  page in order. */
function defaultNext(schema: Schema, id: PageID): PageID | '' {
  const i = schema.pages.findIndex((p) => p.id === id)
  if (i < 0) return ''
  const page = schema.pages[i]!
  if (page.next) return page.next
  return schema.pages[i + 1]?.id ?? ''
}

/**
 * evalCondition resolves one comparison.
 *
 * Exported because the rule builder previews a condition against sample answers
 * before the rule is saved, and because the shared fixture tests it directly.
 *
 * An unanswered field makes every comparison false except the emptiness checks.
 * Treating a missing answer as a zero value would route respondents down the
 * "rating is 0, so ask what went wrong" branch before they had rated anything.
 */
export function evalCondition(c: Condition, answers: Answers): boolean {
  const raw = answers[c.field]
  const answered = c.field in answers && !isEmpty(raw)

  switch (c.op) {
    case 'is_empty':
      return !answered
    case 'is_not_empty':
      return answered
  }
  if (!answered) return false

  switch (c.op) {
    case 'eq':
      return equal(raw, c.value)
    case 'neq':
      return !equal(raw, c.value)
    case 'in':
      return inList(raw, c.value)
    case 'not_in':
      return !inList(raw, c.value)
    case 'contains':
      return contains(raw, c.value)
    case 'gt':
    case 'gte':
    case 'lt':
    case 'lte':
      return compareOrder(c.op, raw, c.value)
    case 'between': {
      const bounds = c.value
      if (!Array.isArray(bounds) || bounds.length !== 2) {
        throw new EngineError('operator "between" needs a two-element value')
      }
      return compareOrder('gte', raw, bounds[0]) && compareOrder('lte', raw, bounds[1])
    }
    default:
      throw new EngineError(`unknown operator "${c.op}"`)
  }
}

function isEmpty(v: unknown): boolean {
  if (v === null || v === undefined) return true
  if (typeof v === 'string') return v.trim() === ''
  if (Array.isArray(v)) return v.length === 0
  if (typeof v === 'object') return Object.keys(v as object).length === 0
  return false
}

/**
 * equal compares an answer against a rule value.
 *
 * Both sides arrive from JSON, where a choice is an option id string and a
 * rating is a number, so comparison normalises to string except when both sides
 * are genuinely numeric.
 */
function equal(answer: unknown, want: unknown): boolean {
  const an = asNumber(answer)
  const wn = asNumber(want)
  if (an !== null && wn !== null) return an === wn
  return asString(answer) === asString(want)
}

function inList(answer: unknown, want: unknown): boolean {
  if (!Array.isArray(want)) return false
  return want.some((item) => equal(answer, item))
}

/** contains checks membership for multi-select answers and substring for text. */
function contains(answer: unknown, want: unknown): boolean {
  if (Array.isArray(answer)) return answer.some((item) => equal(item, want))
  return asString(answer).includes(asString(want))
}

function compareOrder(op: string, answer: unknown, want: unknown): boolean {
  const an = asNumber(answer)
  const wn = asNumber(want)
  if (an !== null && wn !== null) return applyOrder(op, cmpNumber(an, wn))

  // Dates are ISO-8601, which sorts correctly as text, so ordering works for
  // date fields without parsing them.
  const as = asString(answer)
  const ws = asString(want)
  if (as === '' || ws === '') return false
  return applyOrder(op, as < ws ? -1 : as > ws ? 1 : 0)
}

function applyOrder(op: string, c: number): boolean {
  switch (op) {
    case 'gt':
      return c > 0
    case 'gte':
      return c >= 0
    case 'lt':
      return c < 0
    case 'lte':
      return c <= 0
    default:
      return false
  }
}

function cmpNumber(a: number, b: number): number {
  return a < b ? -1 : a > b ? 1 : 0
}

/** asNumber mirrors the Go side, which also parses numeric strings -- a rating
 *  posted as "4" by a form control must compare like the number 4. */
function asNumber(v: unknown): number | null {
  if (typeof v === 'number') return Number.isFinite(v) ? v : null
  if (typeof v === 'string') {
    const t = v.trim()
    if (t === '') return null
    const n = Number(t)
    return Number.isFinite(n) ? n : null
  }
  return null
}

function asString(v: unknown): string {
  if (v === null || v === undefined) return ''
  if (typeof v === 'string') return v
  if (typeof v === 'boolean') return v ? 'true' : 'false'
  if (typeof v === 'number') return String(v)
  return String(v)
}

File: internal/modules/forms/engine/engine.go
// Package engine evaluates a form's conditional logic.
//
// Evaluate is a pure function: no I/O, no clock, no randomness. That is what
// lets the same logic compile to WebAssembly for the browser while the server
// keeps the authoritative copy, with one shared set of golden fixtures keeping
// the two from drifting apart.
//
// The server must always re-evaluate. Trusting the client's claim about which
// fields it displayed would let a caller submit answers for a branch they never
// saw -- and therefore never saw the consent text for.
package engine

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"github.com/collectr/collectr/internal/modules/forms/domain"
)

// Result is the outcome of evaluating a schema against a set of answers.
type Result struct {
	// Visible lists the fields actually shown, in page then declaration order.
	Visible []domain.FieldID
	// Required lists the visible fields that must be answered.
	Required []domain.FieldID
	// Path lists the pages traversed, in order.
	Path []domain.PageID
}

// IsVisible reports whether the field was shown.
func (r Result) IsVisible(id domain.FieldID) bool {
	for _, f := range r.Visible {
		if f == id {
			return true
		}
	}
	return false
}

// Evaluate walks the form as the respondent's answers direct, and reports which
// fields they were actually shown.
func Evaluate(s domain.Schema, answers map[string]any) (Result, error) {
	if len(s.Pages) == 0 {
		return Result{}, nil
	}

	// Rules are grouped by page once, so the walk stays linear rather than
	// rescanning every rule for every page.
	byPage := make(map[domain.PageID][]domain.Rule, len(s.Pages))
	for _, r := range s.Rules {
		byPage[r.OnPage] = append(byPage[r.OnPage], r)
	}

	var (
		res     Result
		current = s.Pages[0].ID
		seen    = make(map[domain.PageID]struct{}, len(s.Pages))
	)

	// Publishing rejects cyclic navigation, but a schema written before that
	// check existed -- or one edited straight in the database -- must not be able
	// to hang a request. The visited set bounds the walk regardless.
	for range len(s.Pages) + 1 {
		page, ok := s.Page(current)
		if !ok {
			return Result{}, fmt.Errorf("page %q referenced but not defined", current)
		}
		if _, repeat := seen[current]; repeat {
			return Result{}, fmt.Errorf("navigation cycle at page %q", current)
		}
		seen[current] = struct{}{}
		res.Path = append(res.Path, current)

		visible, required, next, err := evalPage(s, page, byPage[current], answers)
		if err != nil {
			return Result{}, err
		}
		res.Visible = append(res.Visible, visible...)
		res.Required = append(res.Required, required...)

		if next == "" {
			return res, nil
		}
		current = next
	}
	return Result{}, fmt.Errorf("navigation exceeded %d pages", len(s.Pages))
}

// evalPage resolves one page: which of its fields are shown, which are required,
// and where to go next. next == "" means the form ends here.
func evalPage(s domain.Schema, page domain.Page, rules []domain.Rule, answers map[string]any) (visible, required []domain.FieldID, next domain.PageID, err error) {
	// Start from the declared state, then let rules amend it. A field marked
	// hidden appears only if a rule shows it.
	shown := make(map[domain.FieldID]bool, len(page.Fields))
	must := make(map[domain.FieldID]bool, len(page.Fields))
	for _, fid := range page.Fields {
		f, ok := s.Fields[fid]
		if !ok {
			return nil, nil, "", fmt.Errorf("page %q lists undefined field %q", page.ID, fid)
		}
		shown[fid] = !f.Hidden
		must[fid] = f.Required
	}

	var (
		jump    domain.PageID
		ended   bool
		decided bool
	)

	// One pass, in declaration order. Later rules override earlier ones, which
	// makes the outcome depend only on the schema's own ordering -- no fixpoint
	// iteration, no dependence on map order.
	for _, rule := range rules {
		match, err := evalCondition(rule.When, answers)
		if err != nil {
			return nil, nil, "", fmt.Errorf("rule %q: %w", rule.ID, err)
		}
		actions := rule.Else
		if match {
			actions = rule.Then
		}
		for _, a := range actions {
			switch a.Action {
			case domain.ActionShow:
				shown[domain.FieldID(a.Target)] = true
			case domain.ActionHide:
				shown[domain.FieldID(a.Target)] = false
			case domain.ActionRequire:
				must[domain.FieldID(a.Target)] = true
			case domain.ActionOptional:
				must[domain.FieldID(a.Target)] = false
			case domain.ActionGoto:
				jump, ended, decided = domain.PageID(a.Target), false, true
			case domain.ActionEnd:
				jump, ended, decided = "", true, true
			default:
				return nil, nil, "", fmt.Errorf("rule %q: unknown action %q", rule.ID, a.Action)
			}
		}
	}

	for _, fid := range page.Fields {
		if !shown[fid] {
			continue
		}
		visible = append(visible, fid)
		if must[fid] {
			required = append(required, fid)
		}
	}

	switch {
	case decided && ended:
		return visible, required, "", nil
	case decided:
		return visible, required, jump, nil
	default:
		return visible, required, defaultNext(s, page.ID), nil
	}
}

// defaultNext returns where the form goes when no rule decides: the page's
// declared Next, or the following page in order.
func defaultNext(s domain.Schema, id domain.PageID) domain.PageID {
	for i, p := range s.Pages {
		if p.ID != id {
			continue
		}
		if p.Next != "" {
			return p.Next
		}
		if i+1 < len(s.Pages) {
			return s.Pages[i+1].ID
		}
		return ""
	}
	return ""
}

// evalCondition resolves one comparison.
//
// An unanswered field makes every comparison false except the emptiness checks.
// The alternative -- treating a missing answer as a zero value -- would silently
// route respondents down the "rating is 0, so ask what went wrong" branch before
// they had rated anything.
func evalCondition(c domain.Condition, answers map[string]any) (bool, error) {
	raw, answered := answers[string(c.Field)]
	if answered {
		answered = !isEmpty(raw)
	}

	switch c.Op {
	case domain.OpIsEmpty:
		return !answered, nil
	case domain.OpIsNotEmpty:
		return answered, nil
	}
	if !answered {
		return false, nil
	}

	switch c.Op {
	case domain.OpEq:
		return equal(raw, c.Value), nil
	case domain.OpNeq:
		return !equal(raw, c.Value), nil
	case domain.OpIn:
		return inList(raw, c.Value), nil
	case domain.OpNotIn:
		return !inList(raw, c.Value), nil
	case domain.OpContains:
		return contains(raw, c.Value), nil
	case domain.OpGt, domain.OpGte, domain.OpLt, domain.OpLte:
		return compareOrder(c.Op, raw, c.Value)
	case domain.OpBetween:
		bounds, ok := c.Value.([]any)
		if !ok || len(bounds) != 2 {
			return false, fmt.Errorf("operator %q needs a two-element value", domain.OpBetween)
		}
		lo, err := compareOrder(domain.OpGte, raw, bounds[0])
		if err != nil {
			return false, err
		}
		hi, err := compareOrder(domain.OpLte, raw, bounds[1])
		if err != nil {
			return false, err
		}
		return lo && hi, nil
	default:
		return false, fmt.Errorf("unknown operator %q", c.Op)
	}
}

func isEmpty(v any) bool {
	switch t := v.(type) {
	case nil:
		return true
	case string:
		return strings.TrimSpace(t) == ""
	case []any:
		return len(t) == 0
	case []string:
		return len(t) == 0
	case map[string]any:
		return len(t) == 0
	default:
		return false
	}
}

// equal compares an answer against a rule value.
//
// Both sides arrive from JSON, where every number is a float64 and every choice
// is an option id string, so comparison normalises to string except when both
// sides are genuinely numeric.
func equal(answer, want any) bool {
	if an, ok := asNumber(answer); ok {
		if wn, ok := asNumber(want); ok {
			return an == wn
		}
	}
	return asString(answer) == asString(want)
}

func inList(answer, want any) bool {
	list, ok := want.([]any)
	if !ok {
		return false
	}
	for _, item := range list {
		if equal(answer, item) {
			return true
		}
	}
	return false
}

// contains checks membership for multi-select answers and substring for text.
func contains(answer, want any) bool {
	switch t := answer.(type) {
	case []any:
		for _, item := range t {
			if equal(item, want) {
				return true
			}
		}
		return false
	case []string:
		for _, item := range t {
			if equal(item, want) {
				return true
			}
		}
		return false
	default:
		return strings.Contains(asString(answer), asString(want))
	}
}

func compareOrder(op string, answer, want any) (bool, error) {
	an, aok := asNumber(answer)
	wn, wok := asNumber(want)
	if aok && wok {
		return applyOrder(op, cmpFloat(an, wn)), nil
	}
	// Dates are ISO-8601, which sorts correctly as text, so ordering still works
	// for date fields without parsing them.
	as, ws := asString(answer), asString(want)
	if as == "" || ws == "" {
		return false, nil
	}
	return applyOrder(op, strings.Compare(as, ws)), nil
}

func applyOrder(op string, c int) bool {
	switch op {
	case domain.OpGt:
		return c > 0
	case domain.OpGte:
		return c >= 0
	case domain.OpLt:
		return c < 0
	case domain.OpLte:
		return c <= 0
	default:
		return false
	}
}

func cmpFloat(a, b float64) int {
	switch {
	case a < b:
		return -1
	case a > b:
		return 1
	default:
		return 0
	}
}

func asNumber(v any) (float64, bool) {
	switch t := v.(type) {
	case float64:
		return t, true
	case float32:
		return float64(t), true
	case int:
		return float64(t), true
	case int64:
		return float64(t), true
	case json.Number:
		f, err := t.Float64()
		return f, err == nil
	case string:
		f, err := strconv.ParseFloat(strings.TrimSpace(t), 64)
		return f, err == nil
	default:
		return 0, false
	}
}

func asString(v any) string {
	switch t := v.(type) {
	case nil:
		return ""
	case string:
		return t
	case bool:
		return strconv.FormatBool(t)
	case float64:
		return strconv.FormatFloat(t, 'f', -1, 64)
	default:
		return fmt.Sprint(t)
	}
}
